### PR TITLE
Mobile menu

### LINKED ISSUE #436: Create a sidebar navigation menu for mobile views
I was thinking about a sidebar menu similar to this (apologies in advance mocking is not my strong suit):

image

Some things to considerate:

- Each menu item is collapsible and should display the subitems upon clicking or tapping (check [this](https://github.com/Blobscan/blobscan/blob/next/apps/web/src/components/Collapsable.tsx) and [this](https://github.com/Blobscan/blobscan/blob/next/apps/web/src/components/Rotable.tsx)). 
- The sidebar menu doesn’t open entirely, leaving about 20% of the screen with a smoothly displayed greyish background. Clicking on that background closes the menu smoothly.
- When selecting one of the items and reopening the sidebar menu, the item and subitem where we are located should display a background.

Something similar to this:

image

Check [Blockscout](https://optimism.blockscout.com) for inspiration

### PR DESCRIPTION
**Pull request by luis-herasme**
### Checklist

- [ ] My change requires a documentation update, and I have done it.
- [ ] I have added tests to cover my changes.
- [x] I have filled out the description and linked the related issues.

### Description
1. Adds `MobileMenu` component.
2. Refactored the menu data out of the menu so that the normal menu and the mobile menu can share the same data.
3. Fixes a small error inside the button component of the icon variant, causing the icon to have a width of zero

### Related Issue
Closes #436

### Screenshots:
image

image

Update:

https://github.com/user-attachments/assets/342483b3-f5ef-4661-ac78-c40d666f0f6c

### CODE REVIEW

**Review by PJColombo — CHANGES_REQUESTED** [review]
Looking good!

There are some ideas I had in mind for this. I forgot to add the necessary context to this issue :/.

**Comment by PJColombo** [comment]
I was thinking about a sidebar menu similar to this (apologies in advance mocking is not my strong suit):

image

Some things to considerate:

- Each menu item is collapsible and should display the subitems upon clicking or tapping (check [this](https://github.com/Blobscan/blobscan/blob/next/apps/web/src/components/Collapsable.tsx) and [this](https://github.com/Blobscan/blobscan/blob/next/apps/web/src/components/Rotable.tsx)). 
- The sidebar menu doesn’t open entirely, leaving about 20% of the screen with a smoothly displayed greyish background. Clicking on that background closes the menu smoothly.
- When selecting one of the items and reopening the sidebar menu, the item and subitem where we are located should display a background.


Something similar to this:

image

**Comment by PJColombo** [comment]
Another issue: the menu is not showing in the top bar for any section other than the homepage.

image

**Comment by PJColombo** [comment]
I updated the [issue](https://github.com/Blobscan/blobscan/issues/436) with the design that I had in mind

File: apps/web/src/components/NavMenusSection/index.tsx
Comment (PJColombo) [inline apps/web/src/components/NavMenusSection/index.tsx:1]: Let's rename `NavMenusSection` to `NavMenus`

File: apps/web/src/components/NavMenusSection/data.tsx
Comment (PJColombo) [inline apps/web/src/components/NavMenusSection/data.tsx:1]: Let's move this to a file `content.ts` located at the root of the `/src` folder

I want to improve code discoverability by having a single file containing all static content for the website (links, navigation items, etc).

We'll place here the footer content in a different PR

File: apps/web/src/components/NavMenusSection/data.tsx [hunk 1]
```diff
@@ -0,0 +1,102 @@
...
+
+export type ExpandibleMenuItem = {
+  label: string;
+  icon: ReactNode;
+  items: {
+    label: string;
+    href: string;
+  }[];
+  type: "expandible";
+};
```
Comment (PJColombo) [inline apps/web/src/components/NavMenusSection/data.tsx:43]: Let's rename this types to `NavItem` and `ExpandibleNavItem`

File: apps/web/src/components/NavMenusSection/data.tsx [hunk 2]
```diff
@@ -0,0 +1,102 @@
...
+
+export const MENU_DATA: Array<MenuItem | ExpandibleMenuItem> = [
```
Comment (PJColombo) [inline apps/web/src/components/NavMenusSection/data.tsx:45]: Let's rename this to `MENU_ITEMS`

File: apps/web/src/components/NavMenusSection/data.tsx (hunk 1 above)
Comment (PJColombo) [reply apps/web/src/components/NavMenusSection/data.tsx:43]: We're already using types similar to this one in the `MobileMenu` and  `NavItem`. Let's reuse these ones there

File: apps/web/src/components/NavMenusSection/data.tsx [hunk 3]
```diff
@@ -0,0 +1,102 @@
...
+  }
+
+  return `https://api.${env.NEXT_PUBLIC_NETWORK_NAME}.blobscan.com`;
+}
+
+export type MenuItem = {
+  label: string;
+  href: string;
+  icon: ReactNode;
+  type: "single";
```
Comment (PJColombo) [inline apps/web/src/components/NavMenusSection/data.tsx:32]: No need to have a `type` field. Let's use [type guards](https://www.typescriptlang.org/docs/handbook/advanced-types.html) for each type instead

Comment (PJColombo) [reply apps/web/src/components/NavMenusSection/data.tsx:32]: We can export these functions and use them to determine if an object belongs to one of these types.

File: apps/web/src/components/NavMenusSection/MobileMenu.tsx [hunk 4]
```diff
@@ -0,0 +1,159 @@
...
+    } else {
+      document.body.style.overflow = "auto";
+    }
+  }, [open]);
+
+  return (
+    <>
+      <Button
+        variant="icon"
+        className="md:hidden"
```
Comment (PJColombo) [inline apps/web/src/components/NavMenusSection/MobileMenu.tsx:29]:
```suggestion
```

I prefer to delegate the responsibility of hiding or displaying a component to the component’s consumer.

Comment (PJColombo) [reply apps/web/src/components/NavMenusSection/MobileMenu.tsx:29]: When using `MobileMenu` we'll wrap it within a `div` and add this responsive tailwind classes

**Review by PJColombo — CHANGES_REQUESTED** [review]
Nice job!

It’s starting to take shape nicely.

I’ve left some comments regarding the refactoring of the current components.

File: apps/web/src/components/NavMenusSection/MobileMenu.tsx [hunk 5]
```diff
@@ -0,0 +1,159 @@
...
+      >
+        <ThemeModeButton />
+      </div>
+    </>
+  );
+}
+
+function MobileMenuBackground({
+  open,
+  close,
```
[the remark this answers is not in the document]
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/MobileMenu.tsx]: Done [5c98023](https://github.com/Blobscan/blobscan/pull/469/commits/5c980231ac13efaddef72f79b7b5235f5fba3eef)

File: apps/web/src/components/NavMenusSection/MobileMenu.tsx [hunk 6]
```diff
@@ -0,0 +1,159 @@
...
+        }`}
+      >
+        <ThemeModeButton />
+      </div>
+    </>
+  );
+}
+
+function MobileMenuBackground({
+  open,
```
[the remark this answers is not in the document]
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/MobileMenu.tsx]: Done [1411ead](https://github.com/Blobscan/blobscan/pull/469/commits/1411ead93f9d5aff88698c4d780fad268da59b72)

File: apps/web/src/components/Rotable.tsx [hunk 7]
```diff
@@ -5,7 +5,7 @@ export type RotableProps = {
   children: ReactNode;
   rotated: boolean;
   angle: number;
-  onClick(): void;
+  onClick?: () => void;
```
[the remark this answers is not in the document]
Comment (luis-herasme) [reply apps/web/src/components/Rotable.tsx:8]: Sometimes, it is not necessary to add an onClick. For example, that's the case now in the `MobileMenu`:
https://github.com/Blobscan/blobscan/pull/469/files#diff-f7da5df2e23b948aeb88bc774e6aa80874116724051131e87843d7150261af28R125

File: apps/web/src/components/NavMenusSection/NavItem.tsx
[the remark this answers is not in the document]
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/NavItem.tsx:1]: Done [cce3718](https://github.com/Blobscan/blobscan/pull/469/commits/cce371869d042253680e694d4390e0f29c60371b)

File: apps/web/src/components/NavMenusSection/MobileMenu.tsx
[the remark this answers is not in the document]
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/MobileMenu.tsx:1]: Done 25aa9e7

File: apps/web/src/components/NavMenusSection/MobileMenu.tsx [hunk 8]
```diff
@@ -0,0 +1,159 @@
...
+          open ? "translate-x-0" : "translate-x-[-200px]"
+        }`}
+      >
+        <ThemeModeButton />
+      </div>
+    </>
+  );
+}
+
+function MobileMenuBackground({
```
[the remark this answers is not in the document]
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/MobileMenu.tsx]: Done 2686617

File: apps/web/src/components/NavMenusSection/data.tsx
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/data.tsx:1]: Done 5111397

File: apps/web/src/components/NavMenusSection/index.tsx
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/index.tsx:1]: Done [5111397](https://github.com/Blobscan/blobscan/pull/469/commits/51113977227dc7bbb157b88202b2d59ec31a0e06)

File: apps/web/src/components/NavMenusSection/data.tsx (hunk 2 above)
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/data.tsx]: Done [3c39bab](https://github.com/Blobscan/blobscan/pull/469/commits/3c39babd2d9bd6f2262f68223123c7d822b3d0f2)

File: apps/web/src/components/NavMenusSection/data.tsx (hunk 1 above)
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/data.tsx]: Done c724c67 35e39a9 f9786ba

File: apps/web/src/components/NavMenusSection/data.tsx (hunk 3 above)
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/data.tsx]: Done 2f1289c

File: apps/web/src/components/NavMenusSection/MobileMenu.tsx (hunk 4 above)
Comment (luis-herasme) [reply apps/web/src/components/NavMenusSection/MobileMenu.tsx]: Done [d0a163d](https://github.com/Blobscan/blobscan/pull/469/commits/d0a163dbdb8131bc4286b926ac30f8c536bf521a)

File: apps/web/src/components/AppLayout/TopBarLayout/CompactTopBar.tsx [hunk 9]
```diff
@@ -1,16 +1,16 @@
...
 export const CompactTopBar = function () {
   return (
     <>
-      <div className="z-10 sm:hidden">
+      <div className="z-50 sm:hidden">
         <TopBarSurface>
           <div className="flex w-full items-center justify-between">
             <BlobscanLogo className="w-40" />
-            <ThemeModeButton />
+            <MobileMenu />
```
[the remark this answers is not in the document]
Comment (luis-herasme) [reply apps/web/src/components/AppLayout/TopBarLayout/CompactTopBar.tsx]: Done [4cb79f7](https://github.com/Blobscan/blobscan/pull/469/commits/4cb79f78c6ec19d447b266460da08fd29cebd64d)

**Comment by PJColombo** [comment]
Nice job!

I pushed a few commits fixing some extra things. Some things I'd like to point out:

- I made some component renaming changes.
-  I rollbacked the change I requested about separating all `NavItem` components from the `NavigationMenus` and I put them back again there to maintain consistency with the `SidebarNavigationMenu` component, where the navigation components are also kept within the same file.

**Review by PJColombo — APPROVED** [review]
LGTM!. Waiting for another approve so we can merge this

**Review by PabloCastellano — APPROVED** [review]
looks good at preview

**Review by xFJA — APPROVED** [review]

File: apps/web/src/components/NavMenusSection/MobileMenu.tsx [hunk 10]
```diff
@@ -0,0 +1,67 @@
...
+}) {
+  return (
+    <div
+      className="fixed left-0 top-0 z-10 h-full w-full overflow-y-auto bg-background-light p-4 shadow dark:bg-background-dark"
+      style={{
+        transform: open ? "translateX(0)" : "translateX(-100%)",
+        transition: "transform 0.6s",
+      }}
+    >
+      <div>
```
[the remark this answers is not in the document]
Comment (PJColombo) [reply apps/web/src/components/NavMenusSection/MobileMenu.tsx]: This component is actually outdated

File: apps/web/src/components/NavMenusSection/MobileMenu.tsx [hunk 11]
```diff
@@ -0,0 +1,67 @@
...
+        href={item.href}
+      >
+        <div className="h-4 w-4">{item.icon}</div>
+        {item.label}
+      </a>
+    );
+  }
+
+  return (
+    <div>
```
[the remark this answers is not in the document]
Comment (PJColombo) [reply apps/web/src/components/NavMenusSection/MobileMenu.tsx]: This component is actually outdated

**Pull request merged**